core/src/main/scala/io/projectglow/vcf/VCFStreamWriter.scala

```diff
@@ -93,69 +140,17 @@ class VCFStreamWriter(
 
   override def close(): Unit = {
     // Header must be written before closing writer, or else VCF readers will break.
-    if (!headerHasBeenSetOrWritten && writeHeader) {
-      if (providedSampleIds.isEmpty) {
+    if (header == null && writeHeader) {
+      if (inferSampleIds) {
         throw new IllegalStateException(
           "Cannot infer header for empty partition; " +
           "we suggest calling coalesce or repartition to remove empty partitions.")
       }
+      val (sampleIds, _) = sampleIdsMissingOpt.get
```

Comment:
Shouldn't be using the inferred sample IDs if we inferred them?

Reply:
This happens in the case that `header == null`, which only occurs in the case that we did not encounter any rows to write and therefore cannot have inferred any sample IDs.